src/mongo_odm/test/query_builder.cpp

```diff
@@ -0,0 +1,166 @@
+    int x1;
+    int x2;
+    bool y;
+    std::string z;
+    MONGO_ODM_MAKE_KEYS_MODEL(Bar, NVP(w), NVP(x1), NVP(x2), NVP(y), NVP(z))
+
+    Bar(int64_t w, int x1, int x2, bool y, std::string z) : w(w), x1(x1), x2(x2), y(y), z(z) {
+        _id = bsoncxx::oid{bsoncxx::oid::init_tag_t{}};
+    }
+
+    // default constructor
+    Bar() {
+        _id = bsoncxx::oid{bsoncxx::oid::init_tag_t{}};
+    }
+
+    bsoncxx::oid getID() {
+        return _id;
+    }
+};
+
+// Test ODM on class that does not rely on model
+class Point {
+   public:
+    int x;
+    int y;
+    MONGO_ODM_MAKE_KEYS(Point, NVP(x), NVP(y));
+};
+
+TEST_CASE("Query Builder", "[mongo_odm::query_builder]") {
+    instance::current();
+    client conn{uri{}};
+    collection coll = conn["testdb"]["testcollection"];
+    coll.delete_many({});
+
+    Bar::setCollection(coll);
+    Bar(444, 1, 2, false, "hello").save();
+    Bar(444, 1, 3, false, "hello").save();
+    Bar(555, 10, 2, true, "goodbye").save();
+
+    SECTION("Test == comparison.", "[mongo_odm::ComparisonExpr]") {
+        auto res = Bar::find_one(MONGO_ODM_KEY(Bar::x1) == 1);
+        REQUIRE(res);
+        REQUIRE(res.value().x1 == 1);
+
+        res = Bar::find_one(MONGO_ODM_KEY(Bar::z) == "hello");
+        REQUIRE(res);
+        REQUIRE(res.value().z == "hello");
+    }
+
+    SECTION("Test > comparison.", "[mongo_odm::ComparisonExpr]") {
+        auto res = Bar::find_one(MONGO_ODM_KEY(Bar::x1) > 1);
+        REQUIRE(res);
+        REQUIRE(res.value().x1 > 1);
+    }
+
+    SECTION("Test >= comparison.", "[mongo_odm::ComparisonExpr]") {
+        auto res = Bar::find_one(MONGO_ODM_KEY(Bar::x1) >= 10);
+        REQUIRE(res);
+        REQUIRE(res.value().x1 >= 10);
+    }
+
+    SECTION("Test < comparison.", "[mongo_odm::ComparisonExpr]") {
+        auto res = Bar::find_one(MONGO_ODM_KEY(Bar::x1) < 10);
+        REQUIRE(res);
+        REQUIRE(res.value().x1 < 10);
+    }
+
+    SECTION("Test <= comparison.", "[mongo_odm::ComparisonExpr]") {
+        auto res = Bar::find_one(MONGO_ODM_KEY(Bar::x1) <= 1);
+        REQUIRE(res);
+        REQUIRE(res.value().x1 <= 1);
+    }
+
+    SECTION("Test != comparison.", "[mongo_odm::ComparisonExpr]") {
+        auto res = Bar::find_one(MONGO_ODM_KEY(Bar::x1) != 1);
+        REQUIRE(res);
+        REQUIRE(res.value().x1 != 1);
+
+        res = Bar::find_one(MONGO_ODM_KEY(Bar::z) != "hello");
+        REQUIRE(res);
+        REQUIRE(res.value().z == "goodbye");
+    }
+
+    SECTION("Test $not expression, with operator!.", "[mongo_odm::NotExpr]") {
+        auto res = Bar::find_one(!(MONGO_ODM_KEY(Bar::x1) < 10));
+        REQUIRE(res);
+        REQUIRE(res.value().x1 >= 10);
+
+        res = Bar::find_one(!(MONGO_ODM_KEY(Bar::z) == "hello"));
+        REQUIRE(res);
+        REQUIRE(res.value().z == "goodbye");
+    }
+
+    SECTION("Test expression list.", "[mongo_odm::ExpressionList]") {
+        auto res = Bar::find_one((MONGO_ODM_KEY(Bar::x1) == 1, MONGO_ODM_KEY(Bar::x2) == 2,
```

Comment:
&& uses the `$and` operator, which apparently sometimes differs from a list of operators:
"MongoDB provides an implicit AND operation when specifying a comma separated list of expressions. Using an explicit AND with the $and operator is necessary when the same field or operator has to be specified in multiple expressions."